### PR TITLE
Add API for watch list

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it: 
Contains API changes required for implementing [watch-list KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list)

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
xref: https://github.com/kubernetes/enhancements/pull/3667
#### Does this PR introduce a user-facing change?

```release-note
Introduce API for streaming.

Add SendInitialEvents field to the ListOptions. When the new option is set together with watch=true, it begins the watch stream with synthetic init events followed by a synthetic "Bookmark" after which the server continues streaming events.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 1]
```diff
@@ -66,6 +66,32 @@ type ListOptions struct {
...
+	//   is send and watch stream is continued to be send.
+	//   If RV is unset, this is interpreted as "consistent read" and the
+	//   bookmark event is send when the state is synced at least to the moment
+	//   when request started being processed.
+	// - ResourceVersionMatch = Exact
+	//   Unsupported error is returned.
+	// - ResourceVersionMatch unset (or set to any other value)
+	//   BadRequest error is returned.
+	//
+	// Defaults to true if ResourceVersion="" or ResourceVersion="0" (for backward
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:92]: does anyone know where would be the best place to add defaulting ?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 2]
```diff
@@ -66,6 +66,32 @@ type ListOptions struct {
...
+	//   becomes synced to a resource version as fresh as the one provided by
+	//   the ResourceVersion option. At this point, a synthetic bookmark event
+	//   is send and watch stream is continued to be send.
+	//   If RV is unset, this is interpreted as "consistent read" and the
+	//   bookmark event is send when the state is synced at least to the moment
+	//   when request started being processed.
+	// - ResourceVersionMatch = Exact
+	//   Unsupported error is returned.
+	// - ResourceVersionMatch unset (or set to any other value)
+	//   BadRequest error is returned.
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:90]: `ValidateListOptions` method returns `errors.NewInvalid` which sets `HTTP 422`. If that is okay with you I'll reword this comment.

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:90]: I'm fine with this, but I will leave it to API reviewers - @lavalamp I guess for this case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:92]: We don't yet have a place for it - it's probably best to do the defaulting here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L200

But we should define DefaultListOptions function, and given in the above we convert directly to internalversion, probably here:
https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 3]
```diff
@@ -845,6 +845,9 @@ func growSlice(v reflect.Value, maxCapacity int, sizes ...int) {
 
 // Watch implements storage.Interface.Watch.
 func (s *store) Watch(ctx context.Context, key string, opts storage.ListOptions) (watch.Interface, error) {
+	if opts.SendInitialEvents {
+		return nil, apierrors.NewBadRequest("sendInitialEvents for watch is unsupported by an etcd cluster")
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:849]: Yes - let's make it consistent.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 4]
```diff
@@ -66,6 +66,32 @@ type ListOptions struct {
 	// it does not recognize and will return a 410 error if the token can no longer be used because
 	// it has expired.
 	Continue string
+
+	// SendInitialEvents, when set together with Watch option,
+	// begin the watch stream with synthetic init events to build the
+	// whole state of all resources followed by a synthetic "Bookmark"
+	// event containing a ResourceVersion after which the server
+	// continues streaming events.
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:74]:
```suggestion
	// `sendInitialEvents=true` may be set together with `watch=true`.
	// In that case, the watch stream will begin with synthetic events to
	// produce the current state of objects in the collection. Once all such
	// events have been sent, a synthetic "Bookmark" event  will be sent.
	// The bookmark will report the ResourceVersion (RV) corresponding to the
	// set of objects, and be marked with <annotation we agreed on in the KEP>.
	// Afterwards, the watch stream will proceed as usual, sending watch events
	// corresponding to changes (subsequent to the RV) to objects watched.
```

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:74]: Sorry, I meant to leave this comment on the external version! Good to fix it both places though.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 5]
```diff
@@ -400,6 +400,33 @@ type ListOptions struct {
...
 	// resourceVersion value returned by the server and not miss any modifications.
 	Continue string `json:"continue,omitempty" protobuf:"bytes,8,opt,name=continue"`
+
+	// SendInitialEvents, when set together with Watch option,
+	// begin the watch stream with synthetic init events to build the
+	// whole state of all resources followed by a synthetic "Bookmark"
+	// event containing a ResourceVersion after which the server
+	// continues streaming events.
+	//
+	// When SendInitialEvents option is set, we require ResourceVersionMatch
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:410]:
```suggestion
	// When `sendInitialEvents` option is set, we require `resourceVersionMatch`
```

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 6]
```diff
@@ -400,6 +400,33 @@ type ListOptions struct {
...
+	// option to also be set. The semantic of the watch request is as following:
+	// - ResourceVersionMatch = NotOlderThan
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:412]:
```suggestion
	// - `resourceVersionMatch` = NotOlderThan
```

(multiple instances below)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 7]
```diff
@@ -400,6 +400,33 @@ type ListOptions struct {
...
+	//   It starts with sending initial events for all objects (at some resource
+	//   version), potentially followed by an event stream until the state
+	//   becomes synced to a resource version as fresh as the one provided by
+	//   the ResourceVersion option. At this point, a synthetic bookmark event
+	//   is send and watch stream is continued to be send.
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:417]: Can you just describe the differences from the scenario in the beginning part of this comment?

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:417]: Also please use ``` `resourceVersion` ``` when referring to the field.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 8]
```diff
@@ -400,6 +400,33 @@ type ListOptions struct {
...
+	//   If RV is unset, this is interpreted as "consistent read" and the
+	//   bookmark event is send when the state is synced at least to the moment
+	//   when request started being processed.
+	// - ResourceVersionMatch = Exact
+	//   Unsupported error is returned.
+	// - ResourceVersionMatch unset (or set to any other value)
+	//   BadRequest error is returned.
+	//
+	// Defaults to true if ResourceVersion="" or ResourceVersion="0" (for backward
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:426]: Is that really how you used to trigger this behavior? That seems surprising?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:90]: 422 seems OK, especially if we use that elsewhere.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 8 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:426]: Yes - it's unfortunate, but that's the case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 7 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:417]: I could change it to:
```
is interpreted as "data at least as new as the provided resourceVersion" 
and the bookmark event is send when the state is synced 
to a resourceVersion as fresh as the one provided by the ListOptions
```

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:417]: Let's change and Daniel will take another look later on it :)

[I'm fine with this, but I'm too close to it to be fully confident it's really understandable.]

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+limitations under the License.
+*/
+
+package internalversion
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// SetListOptionsDefaults sets defaults on the provided ListOptions if applicable.
+//
+// TODO: once the watch-list fg is always on we could register this function in the scheme (via AddTypeDefaultingFunc).
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:23]: I think this is correct but please double check.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go (hunk 4 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go]: updated the comment, let me know if it is not what we wanted :)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go [hunk 10]
```diff
@@ -198,7 +200,8 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope *RequestSc
 			return
 		}
 
-		if errs := metainternalversionvalidation.ValidateListOptions(&listOptions); len(errs) > 0 {
+		metainternalversion.SetListOptionsDefaults(&listOptions, utilfeature.DefaultFeatureGate.Enabled(features.WatchList))
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:203]: As pointed in one of my comments above, this will eventually not be needed (once we register this function in metainternalversionscheme.ParameterCode

Maybe add a TODO to clean it up.

Same in get below.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 11]
```diff
@@ -42,5 +42,32 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+			return allErrs
+		}
+		if len(match) > 0 && !options.SendInitialEvents {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided"))
+			return allErrs
+		}
+		if !isWatchListFeatureEnabled {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled"))
+		}
+		if !options.AllowWatchBookmarks {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:65]: wouldn't be better to change the defaulting function to also set `options.AllowWatchBookmarks`?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 12]
```diff
@@ -66,6 +66,31 @@ type ListOptions struct {
 	// it does not recognize and will return a 410 error if the token can no longer be used because
 	// it has expired.
 	Continue string
+
+	// `sendInitialEvents=true` may be set together with `watch=true`.
+	// In that case, the watch stream will begin with synthetic events to
+	// produce the current state of objects in the collection. Once all such
+	// events have been sent, a synthetic "Bookmark" event  will be sent.
+	// The bookmark will report the ResourceVersion (RV) corresponding to the
+	// set of objects, and be marked with <annotation we agreed on in the KEP>.
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:75]: oh, right :) - it was advertised as `"k8s.io/initial-events-end": "true"` - I'm okay with that.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go (hunk 9 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:23]: np, I will do.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go (hunk 10 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go:203]: I will leave instructions in the comment of `SetListOptionsDefaults` function.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 13]
```diff
@@ -42,5 +42,32 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+	if options.SendInitialEvents {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for list"))
+	}
+	return allErrs
+}
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if match := options.ResourceVersionMatch; len(match) > 0 || options.SendInitialEvents {
+		if len(match) == 0 && options.SendInitialEvents {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:54]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 14]
```diff
@@ -66,6 +66,31 @@ type ListOptions struct {
...
+	// Afterwards, the watch stream will proceed as usual, sending watch events
+	// corresponding to changes (subsequent to the RV) to objects watched.
+	//
+	// When `sendInitialEvents` option is set, we require `resourceVersionMatch`
+	// option to also be set. The semantic of the watch request is as following:
+	// - `resourceVersionMatch` = NotOlderThan
+	//   is interpreted as "data at least as new as the provided `resourceVersion`"
+	//   and the bookmark event is send when the state is synced
+	//   to a `resourceVersion` as fresh as the one provided by the ListOptions.
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:84]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go (hunk 9 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go]: https://github.com/kubernetes/kubernetes/issues/115478

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+*/
+
+package internalversion
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// SetListOptionsDefaults sets defaults on the provided ListOptions if applicable.
+//
+// TODO: once the watch-list fg is always on we register this function in the scheme (via AddTypeDefaultingFunc).
+// TODO: when the function is registered in the scheme remove all callers of this method.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:24]: nit: you can add:
TODO(#115478): ...

which makes tracking all changes in the code easier (even if we forget updating the bug)

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:64]: We can't do that, because it's backward incompatible.

But even if it would be, we could only do that if it would be of type `*bool` instead of `bool` (which it isn't)
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L347
Otherwise you're not able to distinguish the two cases:
- I didn't set the field (so we can default it)
- I set it to false to explicitly disable that

**Comment by wojtek-t** [comment]
@lavalamp - I had only one remaining comment - this is ready for your next pass now

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go (hunk 15 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:24]: done

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 16]
```diff
@@ -42,5 +42,28 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+	}
+	return allErrs
+}
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if match := options.ResourceVersionMatch; len(match) > 0 || options.SendInitialEvents {
+		if len(match) == 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless resourceVersionMatch is provided"))
+			return allErrs
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:56]: You don't want to return here, because we effectively want to return all validation errors as part of validation. Same below.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 17]
```diff
@@ -23,12 +23,12 @@ import (
 )
 
 // ValidateListOptions returns all validation errors found while validating the ListOptions.
-func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
+func ValidateListOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	if options.Watch {
+		return validateWatchOptions(options, isWatchListFeatureEnabled)
+	}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:29]: There are some validations below (e.g. Continue or the fact that ResourceVersionMatch is not a random string) that we should share chagin between list  and watch.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 18]
```diff
@@ -42,5 +42,28 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided"))
+			return allErrs
+		}
+		if !isWatchListFeatureEnabled {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled"))
+		}
+		if match != metav1.ResourceVersionMatchNotOlderThan {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("resourceVersionMatch"), match, []string{string(metav1.ResourceVersionMatchNotOlderThan)}))
+		}
+	}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:67]: I think this is slightly hard to follow. I would change it to:
```
match := options.ResourceVersionMatch
if len(match) == 0 && options.SendInitialEvents {
   ...
}
if options.SendInitialEvent {
  if !isWatchListFeatureEnabled {
     ...
  }
  if match != metav1.ResourceVersionMatchNotOlderThan {
     ..
  } else {
    if len(match) > 0 {
       ...
    }
  {
}
```

Feel free to disagree, but for me the above is easier to follow than to understand the your code (although I think it's basically the same).

**Review by wojtek-t — COMMENTED** [review]
Only some comments about the validation logic - other than that it LGTM.

And I think it's mostly ready for @lavalamp pass now.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+package internalversion
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// SetListOptionsDefaults sets defaults on the provided ListOptions if applicable.
+//
+// TODO(#115478): once the watch-list fg is always on we register this function in the scheme (via AddTypeDefaultingFunc).
+// TODO(#115478): when the function is registered in the scheme remove all callers of this method.
+func SetListOptionsDefaults(obj *ListOptions, isWatchListFeatureEnabled bool) {
+	if isWatchListFeatureEnabled && obj.Watch && (obj.ResourceVersion == "" || obj.ResourceVersion == "0") && !obj.SendInitialEvents && len(obj.ResourceVersionMatch) == 0 {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:26]: I have just realised that it won't be possible to opt-out for initial events for rv=""/rv=0.

For example `watch=true&allowWatchBookmarks=true&sendInitialEvents=false&resourceVersion=0` will always be defaulted. Should we change `SendInitialEvents` to `*bool` ?

Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:26]: or we change the validation to allow for setting `sendInitialEvents=false&resourceVersionMatch=NotOlderThan`

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 18 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:75]: reworked, see if you like it now.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 17 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:29]: we do check on `ResourceVersionMatch`, `continue` param was missing (added). 

In general I don't think this validation is complete, for example you could start a watch or a list with `continue` param set which doesn't make sense.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 20]
```diff
@@ -42,5 +42,36 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for list"))
+	}
+	return allErrs
+}
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	match := options.ResourceVersionMatch
+	if options.SendInitialEvents {
+		if len(match) == 0 {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:55]: We can check more - if SendInitialEvents is set, then we know that match should exactly be `NotOlderThan`, right?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 21]
```diff
@@ -42,5 +42,36 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+		if !options.SendInitialEvents {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch is forbidden for watch unless sendInitialEvents is provided"))
+			if !isWatchListFeatureEnabled {
+				allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch is forbidden for watch unless the WatchList feature gate is enabled"))
+			}
+		}
+		if match != metav1.ResourceVersionMatchNotOlderThan {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("resourceVersionMatch"), match, []string{string(metav1.ResourceVersionMatchNotOlderThan)}))
+		}
+		if len(options.Continue) > 0 {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:72]: Let's move it outside of this branch - Continue is unexpected for watch under any case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 22]
```diff
@@ -42,5 +42,36 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
 			allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), "resourceVersionMatch \"exact\" is forbidden for resourceVersion \"0\""))
 		}
 	}
+	if options.SendInitialEvents {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for list"))
+	}
+	return allErrs
+}
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:51]: I think I got slightly lost when reviewing this function, so I need an exact table of what we allow
[I'm assuming it's for watch]

- match=""  -> OK    (backward compatibility, typical old case)
- match=NotOlderThan,  SendInitialEvents  -> OK (new case)

For the case of match="" and  RV=0/RV="", SendInitialEvent will get propagated via default and that will change match to NotOlderThan too, so we will be fine.

Is there anything else we allow? I think that's all.
If so, I think that validation should better reflect that.

```
if len(match) == 0 {
   // typical watch 
  if options.SendInitialEvents {
    forbidden
  }
}
if len(match) > 0 {
  // the new watch with SendInitialEvents
  if match != NotOlderThan {
    forbidden
  }
  if !SendInitialEvents {
   forbidden
  }
  if !FG enabled {
   forbidden
  }
}

if len(continue) > 0 {
  forbidden for watch
}
```

That would be much cleaner for me - WDYT?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 21 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:72]: nope, this will make it a backward-incompatible change.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 23]
```diff
@@ -42,5 +42,36 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+	if options.SendInitialEvents {
+		if len(match) == 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless resourceVersionMatch is provided"))
+		}
+		if !isWatchListFeatureEnabled {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless the WatchList feature gate is enabled"))
+		}
+	}
+	if len(match) > 0 {
+		if !options.SendInitialEvents {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:63]: I think it wasn't allowed, i think the error message is wrong. It should be `resourceVersionMatch is forbidden for watch`. Otherwise it won't be backward compatible.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 21 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:72]: OK - you're right.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go (hunk 20 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:55]: This comment still holds - please fix.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:55]: That would require even more branches. Plus deduplication logic. We do check the value of `ResourceVersionMatch` a few lines below.

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:55]: Why - if SendInitialEvents is set, the only option we support is actuall `NotOlderThan` nothing more.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 24]
```diff
@@ -42,5 +44,37 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+
+func validateWatchOptions(options *internalversion.ListOptions, isWatchListFeatureEnabled bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	match := options.ResourceVersionMatch
+	if options.SendInitialEvents != nil {
+		if len(match) == 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for watch unless resourceVersionMatch is provided"))
+		} else {
+			if match != metav1.ResourceVersionMatchNotOlderThan {
+				allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), fmt.Sprintf("sendInitialEvents requires setting resourceVersionMatch to %s", metav1.ResourceVersionMatchNotOlderThan)))
```
Comment (wojtek-t) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:61]: I would return exactly the same error message for the len(match)==0 case before. So basically:

```
if match != metav1.ResourceVersionMatchNotOlderThan {
	allErrs = append(allErrs, field.Forbidden(field.NewPath("resourceVersionMatch"), fmt.Sprintf("sendInitialEvents requires setting resourceVersionMatch to %s", metav1.ResourceVersionMatchNotOlderThan)))
```

instead of 
```
if len(match) ==0 {
   ...
   } else {
  ...
}
```

**Comment by wojtek-t** [comment]
This LGTM - now we need API approval.

/lgtm

@lavalamp - can you please take a look?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+package internalversion
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// SetListOptionsDefaults sets defaults on the provided ListOptions if applicable.
+//
+// TODO(#115478): once the watch-list fg is always on we register this function in the scheme (via AddTypeDefaultingFunc).
+// TODO(#115478): when the function is registered in the scheme remove all callers of this method.
+func SetListOptionsDefaults(obj *ListOptions, isWatchListFeatureEnabled bool) {
+	if isWatchListFeatureEnabled && obj.Watch && (obj.ResourceVersion == "" || obj.ResourceVersion == "0") && obj.SendInitialEvents == nil && len(obj.ResourceVersionMatch) == 0 {
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go:26]: nit: personally I would find something like this a lot easier to understand:

```
if !isWatchListFeatureEnabled { return }

if obj.SendInitialEvents != nil ||  len(obj.ResourceVersionMatch) != 0 { return }

legacy := obj.ResourceVersion == "" || obj.ResourceVersion == "0"

if obj.Watch && legacy { ... }
```

**Comment by lavalamp** [comment]
LGTM aside from optional nit

/approve

In case you want to fix the nit:
/hold

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go (hunk 25 above)
Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go]: done

**Comment by wojtek-t** [comment]
LGTM

**Comment by p0lyn0mial** [comment]
let's hold this PR until we settle on https://github.com/kubernetes/kubernetes/pull/110960 this will avoid rolling back or issuing new PRs.

**Comment by p0lyn0mial** [comment]
just rebased onto the master branch

**Comment by p0lyn0mial** [comment]
/hold cancel

it looks like https://github.com/kubernetes/kubernetes/pull/110960 is ready, let's merge this PR!

**Comment by wojtek-t** [comment]
The implementation PR is ready for merge and tagged:
https://github.com/kubernetes/kubernetes/pull/110960#issuecomment-1451511688

I'm tagging this one so that we merge API, rebase that one and merge immediately after API.

/lgtm

**Comment by wojtek-t** [comment]
Just for posterity - what changed since last rebase was:
- staging/src/k8s.io/apiserver/pkg/features/kube_features.go - rebased due to config
- staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go
- staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

The API itself wasn't touched - hence I tagged it without re-asking for API re-approval.

**Comment by wojtek-t** [comment]
The implementation PR is now also merged: https://github.com/kubernetes/kubernetes/pull/110960

**Comment by liggitt** [comment]
I'm surprised staging/src/k8s.io/api/testdata/v1.26.0/core.v1.ListOptions.after_roundtrip.pb was added in this PR ... the field is a pointer, right, so old data should have been round-tripped without modification

**Review by hashim21223445 — APPROVED** [review]
